## Trees, Forests, SVMs & k-NN
Here are four common machine learning classification models that we will explor in this notebook:
- Decision Tree
- Random Forest
- Support Vector Machine (SVM)
- K-Nearest Neighbors(K-NN)
My objective is to train these models using the same dataset,test them with the same evaluation metric and compare their performance as well as identify important features.

## Import Libraries

In [33]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.calibration import CalibratedClassifierCV
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score 

## Load the dataset and split it

In [3]:
df=pd.read_csv("healthcare-dataset-stroke-data.csv")
df=df.drop("id",axis=1)
df["bmi"]=df["bmi"].fillna((df["bmi"]).mean())
df=pd.get_dummies(df,drop_first=True)
x=df.drop("stroke",axis=1)
y=df["stroke"]
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)

## Decision Tree
The Decision Tree classifier works by making splits of the input space based on certain conditions on the features values that result in the target prediction,the model is interpretable since the rules learned by the model can be easily represented and understood,but when the tree is too deep it can lead to overfitting of the training data.
In this notebook the maximum depth parameter is set to avoid overfitting.

In [ ]:
decision_tree=DecisionTreeClassifier(max_depth=5,random_state=42)
decision_tree.fit(x_train,y_train)
decision_tree_predict=decision_tree.predict(x_test)

## Random Forest
Random Forest is an ensemble learning algorithm of machine learning where a large number of decision trees are used to make predictions.
The process is based on training each tree using a subset of the entire data randomly and the result is obtained using majority voting,the result is usually more accurate and does not cause any overfitting problem compared to the Decision Tree.
 

In [18]:
randeom_forest=RandomForestClassifier(n_estimators=100,random_state=42)
randeom_forest.fit(x_train,y_train)
randeom_forest_predict=randeom_forest.predict(x_test)

## Support Vector Machine (SVM)
SVM is a classification technique that identifies the best possible decision boundaries for differnetating between classes.
The gole of SVM is to maximize the distance between classes ,an approach which usually increase generalizaion performance.


In [ ]:
svm=CalibratedClassifierCV(SVC(),ensemble=False) ## The updated version of model = SVC(kernel="rbf", probability=True)
svm.fit(x_train,y_train)
svm_pred=svm.predict(x_test)

## K-Nearest-Neighbors(K-NN)
classification of an unseen sample is done based on the lables of it is closest neighbors in the training data using K-NN algorithm.
Prediction is achieved through voting among the closet neighbors. 
It is easy to understand and it is prediction time depends on the size of the dataset.

In [14]:
knn=KNeighborsClassifier(n_neighbors=5)
knn.fit(x_train,y_train)
knn_predict=knn.predict(x_test)

Since the Decision Tree,(K-NN),(SVM),Random Forest models has already been trained using the training data ,the next step involves using it to make predictions about the target("stroke") based on the test data(x_test).
The predictions made by the decision  model will  be used to evaluates it is performance.

## Evaluation

In [31]:

def evaluate(name, y_test, y_pred):
    return {
        "Model": name,
        "Accuracy": accuracy_score(y_test, y_pred),
        "Precision": precision_score(y_test, y_pred, zero_division=0),
        "Recall": recall_score(y_test, y_pred, zero_division=0),
        "F1-score": f1_score(y_test, y_pred, zero_division=0)
    }

Once each classifier has been trained the models are evaluated on the same dataset this evaluation involves accuracy,percision,recall and f1 score for classification performance,evaluation based on the same metrics makes a good comparision of all models possible.

## Compare models together


In [34]:
result=[]
result.append(evaluate("Decision_Tree",y_test,decision_tree_predict))
result.append(evaluate("Random_Forest",y_test,randeom_forest_predict))
result.append(evaluate("KNN",y_test,knn_predict))
result.append(evaluate("SVM",y_test,svm_pred))
result_df=pd.DataFrame(result)
print("Evaluation result for each model is:")
result_df


Evaluation result for each model is:


,Model,Accuracy,Precision,Recall,F1-score
0,Decision_Tree,0.936399,0.2,0.016129,0.029851
1,Random_Forest,0.939335,0.0,0.000000,0.000000
2,KNN,0.936399,0.0,0.000000,0.000000
3,SVM,0.939335,0.0,0.000000,0.000000


Performace of all four models is presented in the table of comparison.
Comparision of several models on same train-test data split makes it possible to find out which algorithm performs better in the case.
This implies that there is no such algorithm that can be considered the best one for all machine learning.

#  Hands-On-Lab (Tasks)
## step 1: Train a decision tree, random forest, SVM, and k-NN on the same train/test split from Day 3.

step one is done all models are trained.

## Step 2: Evaluate all four with the same metric (e.g. F1-score) and assemble the results into one comparison table.


In [35]:
print("Evaluation result for each model is:")
result_df

Evaluation result for each model is:


,Model,Accuracy,Precision,Recall,F1-score
0,Decision_Tree,0.936399,0.2,0.016129,0.029851
1,Random_Forest,0.939335,0.0,0.000000,0.000000
2,KNN,0.936399,0.0,0.000000,0.000000
3,SVM,0.939335,0.0,0.000000,0.000000


Accuracy is the overall percentage of correct predictions.
Precision is the number of positive predictions that turned out to be positive.
Recall is the number of positive instances identified.
F1-score is a balance between precision and recall.
## Key Point About Results (Imbalanced Dataset)
High Accuracy Deception: The models exhibit high accuracy (93.6%), but it can be deceiving.
Precision/Recall = Zero or Low: In the case of algorithms such as Random Forest, KNN, and SVM, precision and recall equal zero. It means that the model prefers to predict the majority class (patients without a stroke).

## Step 3: Report the random forest's top feature importances and interpret them.

In [42]:
importance_f=pd.DataFrame({"Feature":x.columns,"Importance":randeom_forest.feature_importances_})
importance_f=importance_f.sort_values("Importance",ascending=False)
importance_f

,Feature,Importance
3,avg_glucose_level,0.273712
0,age,0.239886
4,bmi,0.234227
12,Residence_type_Urban,0.033473
5,gender_Male,0.031986
14,smoking_status_never smoked,0.027297
1,hypertension,0.026749
2,heart_disease,0.023907
9,work_type_Private,0.023801
7,ever_married_Yes,0.021442


## Explanation of feature importance:
avg_glucose_level is the most important feature selected by the Random Forest,another important variable is age and bmi is an important variable that also help in making predictions,these features seem to carry the most useful information for stroke prediction.

## Step 4: Identify the best-performing model for this dataset and explain, in Markdown, why it likely won.


## The best performing model is :
considering the outcomes of evaluation the Decision Tree has shown the best F1-score among all the models under consideration.
the fact that the performance level of this model is rather low it has become the only one able to identify some cases of the positive class all other models have been predicting the dominant class which resulted in zero percision,Recall,F1-score for the positive class.
The reson for such behavior is likely to be an imbalance in the dataset in which there are far fewer cases of strokes than cases of no strokes.